s4/clarity/_internal/factory.py

```diff
@@ -43,32 +43,73 @@ class ElementFactory(object):
 
-    def __init__(self, lims, element_class, batch_flags=None, request_path=None, name_attribute="name"):
+    def __init__(self, lims, element_class):
         """
+        Creates a new factory to provide interface between Clarity LIMS and the client software.
+        The provided `element_class` will be used to convert the XML records into Python objects
+        and back. There are several class attributes that can be set to configure this interface.
+
+        NAME_ATTRIBUTE: This is used if the name of the class differs from the name in the Clarity API.
+
+        REQUEST_PATH: If the class has the REQUEST_PATH attribute then it will
+        be used to override the default path. It is used in relation to the root URI of the Clarity API.
+
+        BATCH_FLAGS: Batch Flags use the BitField enumeration to create a bit field that is used
+        to determine what batch services are provided for this data type. It defaults to None.
+
         :type lims: LIMS
         :type element_class: classobj
-        :type batch_flags: BatchFlags or None
-        :type request_path: str
-        :param request_path: for example, '/configuration/workflows'.
-                             when not specified, uses '/<plural of element name>'.
-        :type name_attribute: str
-        :param name_attribute: if not "name", provide this to adjust behaviour of 'get_by_name'.
         """
 
         self.lims = lims
         self.element_class = element_class
-        self.name_attribute = name_attribute
-        self.batch_flags = batch_flags or BatchFlags.NONE
+        self.name_attribute = self._get_name_attribute()
+        self.batch_flags = self._get_batch_flag()
         self._plural_name = self.element_class.__name__.lower() + "s"
-
-        if request_path is None:
-            request_path = "/" + self._plural_name
-        self.uri = lims.root_uri + request_path
+        self.uri = self._get_endpoint_uri()
 
         self._cache = dict()
 
         lims.factories[element_class] = self
 
+    def _get_name_attribute(self):
+        if hasattr(self.element_class, 'NAME_ATTRIBUTE'):
+            return self.element_class.NAME_ATTRIBUTE
+        return "name"
+
+    def _get_endpoint_uri(self):
+        # type: () -> str
+        """
+        Returns the Clarity endpoint URI the factory will manage.
+        :return: The fully qualified endpoint URI for this data type.
+        """
+
+        # If the class has the REQUEST_PATH attribute then it will
+        # be used to override the default path.
+        if hasattr(self.element_class, 'REQUEST_PATH'):
+            request_path = self.element_class.REQUEST_PATH
+        else:
+            # The default is the plural name, ex:
+            # /api/v2/steps
+            request_path = "/" + self._plural_name
+
+        # Generate our uri with the request path
+        # in relation to the root uri
+        return self.lims.root_uri + request_path
+
+    def _get_batch_flag(self):
```

Comment:
Shouldn't this be _get_batch_flags?

Reply:
Replacing with Mike's suggested inline code.